lib/scan.ts

```diff
@@ -9,13 +9,23 @@ import { getArchivePath, getImageType } from "./image-type";
 import * as staticModule from "./static";
 import { ImageType, PluginOptions, PluginResponse } from "./types";
 
+// Registry credentials may also be provided by env vars. When both are set, flags take precedence.
+export function mergeEnvVarsIntoCredentials(
+  options: Partial<PluginOptions>,
+): void {
+  options.username = options.username || process.env.SNYK_REGISTRY_USERNAME;
+  options.password = options.password || process.env.SNYK_REGISTRY_PASSWORD;
+}
+
 export async function scan(
   options?: Partial<PluginOptions>,
 ): Promise<PluginResponse> {
   if (!options) {
     throw new Error("No plugin options provided");
   }
 
+  mergeEnvVarsIntoCredentials(options);
```

Comment:
Mutating the data is not great as it makes it harder to reason about how values change over time. If you assume they are set only once it makes it much easier to debug!
A suggestion I can give is to make the decision which variable to use closer to the place where it's needed. Right now credentials are needed only in `imageIdentifierAnalysis()`, so maybe instead of passing options.username/options.password here: https://github.com/snyk/snyk-docker-plugin/blob/095a3a020ce8db747739759f7b538f6b862bd7e1/lib/scan.ts#L106-L107 we could do something like this:
```typescript
const username = options.username || process.env.SNYK_REGISTRY_USERNAME;
const password = options.password || process.env.SNYK_REGISTRY_PASSWORD;
const archiveResult = await getImageArchive(
    targetImage,
    imageSavePath,
    username,
    password,
    options.platform,
  );
```

Reply:
Very spooky because these were both points I thought about and discussed in chat with @jmickey yesterday! Totally agree on mutation. My instinct was to use a new var for the resolved fields, but we agreed in the end that the benefits were outweighed by the risk of accidentally referring to the "username" at a later point in the code and forgetting that it may not be the actual username resolved from considering env vars too. We also talked about resolving in the env vars much later in the flow, at the latest possible moment, just like you suggested. Great minds eh! Eventually decided that doing it as near as poss to start of the plugin's scope would be preferable for similar reasons to the above, ie we could refer to "options.username" at any point and know that it would have been resolved and abstracted away from the config sources (flags, env vars) - things that the plugin logic shouldnt need to know about. Would be happy to chat over this again, let me know what you think :)

Reply:
I've realised that next time I should document these sorts of decisions in the PR at the very least, to save us all a bunch of typing!

Reply:
My reasoning for advocating for mutation in this case is that my understanding was that we pass the `options` object further down the stack. There is an argument that having the `PluginOptions` type and not utilising it further down the stack might also cause confusion.

Reply:
Sounds great! Thanks for explaining the reasoning